### PR TITLE
Improve vanishing route line test.

### LINKED ISSUE #3365: Cover the route line styling with automated test.
This ticket should provide the automated test for route line styling effect, including the congestion level displaying in alternative route, the step/fading gradient color on route lines.

### LINKED ISSUE #3607: [Bug]: CarPlay couldn't have soft color gradient
### Mapbox Navigation SDK version

2.0.0-

### Steps to reproduce

1. `carPlayManager.carPlayNavigationViewController?.navigationMapView?.crossfadesCongestionSegments = true`
2. Build `Example-CarPlay` and start navigation

### Expected behavior

The route line gradient shows soft change.

### Actual behavior

The route line gradient transition is sharp.

### Is this a one-time issue or a repeatable issue?

repeatable

### PR DESCRIPTION
**Pull request by ShanMa1991**
### Description
This pr is to fix #3365 and #3607, and also to combine the segments with the different congestion level but same color, to avoid extra gradient stops.

### Implementation
This pr is adding tests to cover the following functions:
- [x] the basic projected distance calculation, 
- [x] the one coordinates of last step of each route leg leading index issue
- [x] the turn on/off of the `routeLineTracksTraversal` during navigation, 
- [x] the turn on/off of the `crossfadesCongestionSegments` during navigation, 
- [x] the distance threshold to update the route line, 
- [x] the initializing of stored properties of route line and routeProgress update. 
- [x] when there's no congestion level information in the route leg, the `trafficUnknownColor` would be applied to the route line.

This pr also fixes #3607 by allowing the navigationMapView regenerate the gradient stops or route line when `navigationMapView.crossfadesCongestionSegments` turns on/off during active navigation. When `routeLineTracksTraversal` enabled, the change of `navigationMapView.crossfadesCongestionSegments` will generate a new line gradient stops. When the next location update comes in, the line gradient will be updated. When `routeLineTracksTraversal` disabled, the change of `navigationMapView.crossfadesCongestionSegments` will generate a new route line.

### Screenshots or Gifs
After the fix, when enable `carPlayManager.carPlayNavigationViewController?.navigationMapView?.crossfadesCongestionSegments` as `true`:

carPlaySoft

During the active navigation, uses the resume button to switch on and off `navigationMapView?.crossfadesCongestionSegments`  , could see the color gradient transition change
switchResume

### CODE REVIEW

**Comment by ShanMa1991** [comment]
Right now the calculation of `routeRemainingDistancesIndex` is blocking the #3319.
To make sure the `routeRemainingDistancesIndex` is correctly calculated and could pass the parity check with Android, the implementation of the `LineString.ineSliceAlong(startDist: stopDist:)` in [Turf-swift Pr](https://github.com/mapbox/turf-swift/pull/166) is needed.

**Comment by ShanMa1991** [comment]
Added the `LineString.ineSliceAlong(startDist: stopDist:)` in the extension of `MapboxCoreNavigation` to support the calculation of `NavigationMapView. routeRemainingDistancesIndex`.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift [hunk 1]
```diff
@@ -238,15 +232,20 @@ extension NavigationMapView {
...
             var minimumSegment: (Double, UIColor) = isMain ? (0.0, .trafficUnknown) : (0.0, .alternativeTrafficUnknown)
 
             for (index, feature) in congestionFeatures.enumerated() {
                 var associatedFeatureColor = routeCasingColor
-                if case let .string(congestionLevel) = feature.properties?[CongestionAttribute],
-                   case let .boolean(isCurrentLeg) = feature.properties?[CurrentLegAttribute],
+                if case let .boolean(isCurrentLeg) = feature.properties?[CurrentLegAttribute],
                    isCurrentLeg {
-                    associatedFeatureColor = congestionColor(for: congestionLevel, isMain: isMain)
+                    if case let .string(congestionLevel) = feature.properties?[CongestionAttribute] {
```
Comment (ShanMa1991) [inline Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:244]: When there's no congestion level information in the feature, we should default apply the `trafficUnknownColor` to the route line.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift [hunk 2]
```diff
@@ -238,15 +232,20 @@ extension NavigationMapView {
                     return nil
                 }
             }.reduce(0, +)
+            // lastRecordSegment records the last segmentEndPercentTraveled and associated congestion color added to the gradientStops.
+            var lastRecordSegment: (Double, UIColor) = (0.0, traversedRouteColor)
```
Comment (ShanMa1991) [inline Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:236]: `lastRecordSegment` is used to record the last congestion color added to the `gradientStops`. If the current congestion color is the same as previous one, they will be combined.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift [hunk 3]
```diff
@@ -35,15 +35,20 @@ extension NavigationMapView {
...
+            let endIndex = routeLeg.steps.count - 1
+            for (index, routeStep) in routeLeg.steps.enumerated() {
                 if let routeShape = routeStep.shape {
-                    if !routeShape.coordinates.isEmpty {
-                        return routeShape.coordinates
-                    } else { return [] }
+                    if index != endIndex {
+                        legCoordinates.append(routeShape.coordinates)
+                    } else if let finalStep = routeShape.coordinates.first {
+                        legCoordinates.append([finalStep])
```
Comment (1ec5) [inline Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:45]: It would be clearer to map the steps to coordinates arrays, as before, but then drop the last coordinate if it’s a duplicate, or perhaps only do that for an arrival step.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    func getRouteProgress() -> RouteProgress {
+        let route = getRoute()
+        let routeProgress = RouteProgress(route: route, options: routeOptions, legIndex: 0, spokenInstructionIndex: 0)
+        routeProgress.currentLegProgress = RouteLegProgress(leg: route.legs[0], stepIndex: 2, spokenInstructionIndex: 0)
+        routeProgress.currentLegProgress.currentStepProgress = RouteStepProgress(step: route.legs[0].steps[2], spokenInstructionIndex: 0)
+        return routeProgress
+    }
+    
+    func linaGradientToString(lineGradient: Value<StyleColor>?) -> String {
+        return lineGradient.debugDescription.components(separatedBy: "(").last!.components(separatedBy: ")").first!
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:66]: Nit: even tough using force unwrapping in tests can be sometimes useful to see errors early, while running in Xcode it leads to undesired stops while running all tests. I'd recommend to remove force unwrapping and add assert, in case of string is invalid.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    func testRouteLineGradientWithCombinedColor() {
+        let route = getRoute()
+        
+        navigationMapView.trafficModerateColor = navigationMapView.trafficUnknownColor
+        navigationMapView.routes = [route]
+        navigationMapView.routeLineTracksTraversal = true
+        navigationMapView.show([route], legIndex: 0)
+        
+        let expectedGradientStops = [0.0 : navigationMapView.trafficUnknownColor]
+        XCTAssertEqual(expectedGradientStops, navigationMapView.currentLineGradientStops, "Failed to combine the same color of congestion sgement.")
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:186]:
```suggestion
        XCTAssertEqual(expectedGradientStops, navigationMapView.currentLineGradientStops, "Failed to combine the same color of congestion segment.")
```

File: Tests/MapboxCoreNavigationTests/Extensions/LineStringTest.swift [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import XCTest
+import Turf
+import CoreLocation
+import TestHelper
+@testable import MapboxCoreNavigation
+
+class LineStringTests: TestCase {
```
Comment (MaximAlien) [inline Tests/MapboxCoreNavigationTests/Extensions/LineStringTest.swift:7]: Should `LineStringTests` be part of Turf instead?

File: Sources/MapboxCoreNavigation/LineString.swift [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import Turf
+
+extension LineString {
+    /**
+     Returns the portion of the line string that begins at the given start distance and extends the given stop distance along the line string.
+     
+     This method is roughly equivalent to the [turf-line-slice-along](https://turfjs.org/docs/#lineSliceAlong) package of Turf.js ([source code](https://github.com/Turfjs/turf/tree/master/packages/turf-line-slice-along/)).
+     */
+    public func lineSliceAlong(startDistance: LocationDistance, stopDistance: LocationDistance) -> LineString? {
```
[the remark this answers is not in the document]
Comment (ShanMa1991) [reply Sources/MapboxCoreNavigation/LineString.swift:9]: Yes, thanks for the suggestion, I'll rebase it and remove this part after we adopted the latest release of [turf-swift](https://github.com/mapbox/turf-swift) that include this method.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift [hunk 8]
```diff
@@ -59,25 +64,29 @@ extension NavigationMapView {
...
         
         /**
          Add to the count of remaining points all of the remaining legs.
          */
-        for index in stride(from: routeProgress.legIndex + 1, to: completeRoutePoints.nestedList.count, by: 1) {
+        let startIndex = routeProgress.legIndex + 1
+        let endIndex = completeRoutePoints.nestedList.endIndex
+        for index in startIndex..<endIndex {
             allRemainingPoints += completeRoutePoints.nestedList[index].flatMap{ $0 }.count
         }
```
[the remark this answers is not in the document]
Comment (ShanMa1991) [reply Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:91]: Thanks for the suggestion, I'll update this part to remove the for loop.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift (hunk 3 above)
Comment (ShanMa1991) [reply Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:45]: Thanks for the suggestion, I'll update this part.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift (hunk 4 above)
Comment (ShanMa1991) [reply Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:66]: Thanks for the suggestion, I'll add the assertion for helping methods in tests.

File: Tests/MapboxCoreNavigationTests/Extensions/LineStringTest.swift (hunk 6 above)
Comment (ShanMa1991) [reply Tests/MapboxCoreNavigationTests/Extensions/LineStringTest.swift:7]: Yes, I'll rebase the pr and remove this part after we adopted the latest release of turf-swift that include this method.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift [hunk 9]
```diff
@@ -37,9 +37,7 @@ extension NavigationMapView {
         let nestedList = route.legs.map { (routeLeg: RouteLeg) -> [[CLLocationCoordinate2D]] in
             return routeLeg.steps.map { (routeStep: RouteStep) -> [CLLocationCoordinate2D] in
                 if let routeShape = routeStep.shape {
-                    if !routeShape.coordinates.isEmpty {
-                        return routeShape.coordinates
-                    } else { return [] }
+                    return routeShape.coordinates
```
Comment (ShanMa1991) [inline Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:40]: Even `mapbox-directions-swift` parses the route with one more duplicate coordinate in the last step of each route leg, we don't want to drop it here. Because in the `navigationMapView.updateUpcomingRoutePointIndex(routeProgress:)`, we still have to deal with the coordinate of the route step. In this case, we still store them all to be synced with the `routeProgress` that comes in later.

**Comment by MaximAlien** [comment]
I've noticed that whenever we drive through the tunnel and when style changes route line will blink without taking into account vanishing point and current progress, for example:

https://user-images.githubusercontent.com/1496498/142293758-e1254001-a0af-4e1f-bd81-0dc759d480e5.mov

We should fix this issue as a tail-work.

**Comment by MaximAlien** [comment]
Can you make sure that `crossfadesCongestionSegments` option works as expected on CarPlay as well? It doesn't seem to work now when we set:
```swift
carPlayManager.carPlayNavigationViewController?.navigationMapView?.crossfadesCongestionSegments = true
```

For example

Screen Shot 2021-11-17 at 2 41 17 PM
:

**Comment by ShanMa1991** [comment]
> Can you make sure that `crossfadesCongestionSegments` option works as expected on CarPlay as well? It doesn't seem to work now when we set:

Yes, sure, in fact we should also re-generate the gradient stops when this value change and display it.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift [hunk 10]
```diff
@@ -59,26 +57,30 @@ extension NavigationMapView {
...
+        let currentLegSteps = completeRoutePoints.nestedList[routeProgress.legIndex]
+        var allRemainingPoints = 0
         /**
          Find the count of remaining points in the current step.
          */
-        var allRemainingPoints = getSlicedLinePointsCount(currentLegProgress: currentLegProgress, currentStepProgress: currentStepProgress)
+        let lineString = currentStepProgress.step.shape ?? LineString([])
+        if currentStepProgress.distanceTraveled < 0 {
+            allRemainingPoints += currentLegSteps[currentLegProgress.stepIndex].count
+        } else if let slicedLineString = lineString.trimmed(from: currentStepProgress.distanceTraveled, to: currentStepProgress.step.distance) {
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:68]: If I understand correctly `LineString.trimmed(from:to:)` was added in Turf `v2.1.0`, but we did not update Mapbox Directions to it. Should we create new version of Mapbox Directions `v2.1.0-rc.2`, which will use Turf `v2.1.0`? It'll allow to drop explicit Turf usage in Mapbox Navigation.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+    
+    func testUpdateUpcomingRoutePointIndex() {
+        // https://github.com/mapbox/mapbox-navigation-android/blob/0ca183f7cb7bec930521ea9bcd59d0e8e2bef165/libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApiTest.kt#L802
+        let route = getRoute()
+        
+        navigationMapView.initPrimaryRoutePoints(route: route)
+        navigationMapView.routeLineGranularDistances = nil
+        XCTAssertEqual(navigationMapView.fractionTraveled, 0.0)
+        
+        let routeProgress = getRouteProgress()
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:106]: It is unclear what this method returns, can you please improve its signature or provide documentation comment? E.g. why stepIndex is set to `2` specifically etc.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+    func testUpdateRouteLineWithDifferentDistance() {
+        let route = getRoute()
+        let routeProgress = getRouteProgress()
+        let coordinate = route.shape!.coordinates[1]
+        
+        navigationMapView.routes = [route]
+        navigationMapView.routeLineTracksTraversal = true
+        navigationMapView.show([route], legIndex: 0)
+        navigationMapView.updateUpcomingRoutePointIndex(routeProgress: routeProgress)
+        setUpCameraZoom(at: 5.0)
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:146]: `testUpdateRouteLineWithDifferentDistance` passes even without setting zoom level, do we need it here? I'd also add brief comment on what exactly you do in each test. At this point it's hard to follow some of them.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+        setUpCameraZoom(at: 16.0)
+        
+        navigationMapView.routes = [route]
+        navigationMapView.routeLineTracksTraversal = true
+        navigationMapView.show([route], legIndex: 0)
+        navigationMapView.updateUpcomingRoutePointIndex(routeProgress: routeProgress)
+        navigationMapView.travelAlongRouteLine(to: coordinate)
+
+        let fractionTrvaled = navigationMapView.fractionTraveled
+        let fractionTrvaledNextDown = Double(CGFloat(fractionTrvaled).nextDown)
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:218]:
```suggestion
        let fractionTraveled = navigationMapView.fractionTraveled
        let fractionTraveledNextDown = Double(CGFloat(fractionTraveled).nextDown)
```

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 14]
```diff
@@ -0,0 +1,262 @@
+import XCTest
+import MapboxDirections
+import TestHelper
+import Turf
+import MapboxMaps
+@testable import MapboxNavigation
+@testable import MapboxCoreNavigation
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:7]: There is no need to directly use Turf:
```suggestion
import XCTest
import MapboxDirections
import TestHelper
import MapboxMaps
@testable import MapboxNavigation
@testable import MapboxCoreNavigation
```

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift (hunk 10 above)
Comment (ShanMa1991) [reply Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:68]: Because right now `MapboxDirections` is not using the new API of `LineString.trimmed(from:to:)` yet of the Turf `v2.1.0`, I'm not sure if we need to upgrade `MapboxDirections` to include this one.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift (hunk 12 above)
Comment (ShanMa1991) [reply Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:146]: Thanks for the suggestion, I'll add the comment and documentation for each test. The default zoom level is smaller than `1` which could pass the test.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift (hunk 14 above)
Comment (ShanMa1991) [reply Tests/MapboxNavigationTests/VanishingRouteLineTests.swift]: Thanks for the suggestion, the extra directly importing is removed.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift (hunk 11 above)
Comment (ShanMa1991) [reply Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:106]: Thanks for the suggestion, comments are added in each unit test for documentation.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift (hunk 10 above)
Comment (1ec5) [reply Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:68]: It’s OK for MapboxNavigation to explicitly depend on Turf if it has a more specific need than MapboxDirections.

**Review by 1ec5 — APPROVED** [review]
Remember to describe the changes that fixed #3607 in the changelog.
